src/Rewards.sol

```diff
@@ -0,0 +1,239 @@
+    string public constant SIGNATURE_VERSION = "1";
+
+    /**
+     * @dev This constant defines the reward type.
+     * This should be kept consistent with the Reward struct.
+     */
+    bytes public constant REWARD_TYPE = "Reward(address recipient,uint256 amount,uint256 sequence)";
+
+    /**
+     * @dev The maximum period for reward quota renewal. This is to prevent overflows while avoiding the ongoing overhead of safe math operations.
+     */
+    uint256 public constant MAX_PERIOD = 30 days;
+
+    /**
+     * @dev Reference to the NODL token contract.
+     */
+    NODL public nodlToken;
+
+    /**
+     * @dev Maximum amount of rewards that can be distributed in a period.
+     */
+    uint256 public rewardQuota;
+
+    /**
+     * @dev Duration of each reward period.
+     */
+    uint256 public rewardPeriod;
+
+    /**
+     * @dev Timestamp indicating when the reward quota is due to be renewed.
+     */
+    uint256 public quotaRenewalTimestamp;
+
+    /**
+     * @dev Amount of rewards claimed in the current period.
+     */
+    uint256 public rewardsClaimed;
+
+    /**
+     * @dev Address of the authorized oracle.
+     */
+    address public authorizedOracle;
+
+    /**
+     * @dev Mapping to store reward sequences for each recipient to prevent replay attacks.
+     */
+    mapping(address => uint256) public rewardSequences;
+
+    /**
+     * @dev Struct on which basis an individual reward must be issued.
+     */
+    struct Reward {
+        address recipient;
+        uint256 amount;
+        uint256 sequence;
+    }
+
+    /**
+     * @dev Error when the reward quota is exceeded.
+     */
+    error RewardQuotaExceeded();
+
+    /**
+     * @dev Error indicating the reward renewal period is set to zero which is not acceptable.
+     */
+    error ZeroPeriod();
+
+    /**
+     * @dev Error indicating that scheduling the reward quota renewal has failed most likley due to the period being too long.
+     */
+    error TooLongPeriod();
+
+    /**
+     * @dev Error when the reward is not from the authorized oracle.
+     */
+    error UnauthorizedOracle();
+
+    /**
+     * @dev Error when the recipient's reward sequence does not match.
+     */
+    error InvalidRecipientSequence();
+
+    /**
+     * @dev Event emitted when the reward quota is set.
+     */
+    event RewardQuotaSet(uint256 quota);
+
+    /**
+     * @dev Event emitted when a reward is minted.
+     */
+    event RewardMinted(address indexed recipient, uint256 amount, uint256 totalRewardsClaimed);
+
+    /**
+     * @dev Initializes the contract with the specified parameters.
+     * @param token Address of the NODL token contract.
+     * @param initialQuota Initial reward quota.
+     * @param initialPeriod Initial reward period.
+     * @param oracleAddress Address of the authorized oracle.
+     */
+    constructor(NODL token, uint256 initialQuota, uint256 initialPeriod, address oracleAddress)
+        EIP712(SIGNING_DOMAIN, SIGNATURE_VERSION)
```

Comment:
I do not think we need to use variables here. In fact, since EIP712 sub-contract has its own internal variables we are storing these values twice instead of once (once in the EIP712 sub-contract and once in our own constants)

Reply:
Not using varaibles here, just constants. I double checked my own understanding that constants are only known at compile time and will be otpimized by the compiler https://learnsolidity.net/understanding-constants-in-solidity/#2_Importance_of_Constants_in_Solidity

Reply:
Isn't marking those as ` public` making solidity generate public accessor functions that would thus prevent this optimization?

Reply:
Apparently it will not impact optimization and gas costs during the execution. But can impact the bytecode size slightly to the getter functions.

Reply:
It's a nice to have feature to be able to check a deoployed contract for the actual values of these essentail constants for the purpsoe of signing for their target. So with no executaion overhead, I lean towards keeping them. Otherwise I can convert them to private.

Reply:
Remix testing shows you are correct, there are no gas differences. We are all good!